go-controller/pkg/util/net_linux.go

```diff
@@ -0,0 +1,128 @@
+// +build linux
+
+package util
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/vishvananda/netlink"
+)
+
+// GetLinkWithStateUp returns the netlink device with its state marked up
+func GetLinkWithStateUp(interfaceName string) (netlink.Link, error) {
+	link, err := netlink.LinkByName(interfaceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup link %s: %v", interfaceName, err)
+	}
+	err = netlink.LinkSetUp(link)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set the link %s up: %v", interfaceName, err)
+	}
+	return link, nil
+}
+
+// LinkAddrAdd removes existing addresses on the link and adds the new address
+func LinkAddrAdd(link netlink.Link, address string) error {
+	addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return fmt.Errorf("failed to list addresses for the link %s: %v", link.Attrs().Name, err)
+	}
+	for _, addr := range addrs {
+		err = netlink.AddrDel(link, &addr)
+		if err != nil {
+			return fmt.Errorf("failed to delete address on link %s: %v", link.Attrs().Name, err)
+		}
+	}
+
+	ipnet, err := netlink.ParseIPNet(address)
+	if err != nil {
+		return fmt.Errorf("failed to parse ip %s :%v\n", address, err)
+	}
+	err = netlink.AddrAdd(link, &netlink.Addr{IPNet: ipnet})
+	if err != nil {
+		return fmt.Errorf("failed to add address %s on link %s: %v", address, link.Attrs().Name, err)
+	}
+	return nil
+}
+
+// LinkRouteAdd flushes any existing routes for given subnet and adds a new route
+// for that subnet through the gwIPstr
+func LinkRouteAdd(link netlink.Link, gwIPstr string, subnets []string) error {
+	gwIP := net.ParseIP(gwIPstr)
+	if gwIP == nil {
+		return fmt.Errorf("gateway IP %s is not a valid IPv4 or IPv6 address", gwIPstr)
+	}
+	family := netlink.FAMILY_V4
+	if gwIP.To4() == nil {
+		family = netlink.FAMILY_V6
+	}
+	for _, subnet := range subnets {
+		dstIPnet, err := netlink.ParseIPNet(subnet)
+		if err != nil {
+			return fmt.Errorf("failed to parse subnet %s :%v\n", subnet, err)
+		}
+		routeFilter := &netlink.Route{Dst: dstIPnet}
+		filterMask := netlink.RT_FILTER_DST
+		routes, err := netlink.RouteListFiltered(family, routeFilter, filterMask)
+		if err != nil {
+			return fmt.Errorf("failed to get the list of routes for subnet %s", subnet)
+		}
+		for _, route := range routes {
+			err = netlink.RouteDel(&route)
+			if err != nil {
+				return fmt.Errorf("failed to delete route for subnet %s : %v\n", subnet, err)
+			}
+		}
+		route := &netlink.Route{
+			Dst:       dstIPnet,
+			LinkIndex: link.Attrs().Index,
+			Scope:     netlink.SCOPE_UNIVERSE,
+			Gw:        gwIP,
+		}
+		err = netlink.RouteAdd(route)
```

Comment:
@girishmg I think you should use RouteReplace() here. RouteAdd() uses `flags := unix.NLM_F_CREATE | unix.NLM_F_EXCL | unix.NLM_F_ACK` which will leave the route untouched if it already exists (because of `NLM_F_EXCL`). Pretty sure we want to force it to be what we want, rather than leaving old info.  RouteReplace() uses `NLM_F_REPLACE`.

Reply:
@dcbw this was more so to simulate the current `ip route flush dev <blah>`. The RouteReplace() will not do that. See my comment below on `AddrReplace()`